src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs

```diff
@@ -0,0 +1,489 @@
+        {
+            this.CommitUtf8JsonWriterContentsToBuffer();
+
+            WriteItemWithSeparatorIfNeeded();
+
+            this.currentTextWriterContentType = contentType;
+            bool isWritingJson = String.IsNullOrEmpty(this.currentTextWriterContentType) || this.currentTextWriterContentType.StartsWith(MimeConstants.MimeApplicationJson, StringComparison.Ordinal);
+
+            // We don't perform escaping when the content type is not JSON
+            // since we assume the input to be already a valid JSON with all escaping handled by the caller.
+            if (!isWritingJson)
+            {
+                this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+            }
+
+            this.Flush();
+
+            this.textWriter = new ODataUtf8JsonTextWriter(this, isWritingJson);
+
+            return this.textWriter;
+        }
+
+        /// <summary>
+        /// Ends a TextWriter value scope.
+        /// </summary>
+        public void EndTextWriterValueScope()
+        {
+            bool isWritingJson = String.IsNullOrEmpty(this.currentTextWriterContentType) || this.currentTextWriterContentType.StartsWith(MimeConstants.MimeApplicationJson, StringComparison.Ordinal);
+
+            if (!isWritingJson)
+            {
+                this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+                this.textWriter.Dispose();
+                this.Flush();
+
+                CheckIfSeparatorNeeded();
+                CheckIfManualValueAtArrayStart();
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously starts a TextWriter value scope with the specified content type.
+        /// </summary>
+        /// <param name="contentType">The content type of the TextWriter value scope.</param>
+        /// <returns>A task representing the asynchronous operation. The task result is a TextWriter for writing JSON data.</returns>
+        public async Task<TextWriter> StartTextWriterValueScopeAsync(string contentType)
+        {
+            await this.FlushAsync().ConfigureAwait(false);
+
+            WriteItemWithSeparatorIfNeeded();
+
+            this.currentTextWriterContentType = contentType;
+
+            bool isWritingJson = String.IsNullOrEmpty(this.currentTextWriterContentType) || this.currentTextWriterContentType.StartsWith(MimeConstants.MimeApplicationJson, StringComparison.Ordinal);
+
+            if (!isWritingJson)
+            {
+                this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+            }
+
+            await this.FlushAsync().ConfigureAwait(false);
+
+            this.textWriter = new ODataUtf8JsonTextWriter(this, isWritingJson);
+
+            return this.textWriter;
+        }
+
+        /// <summary>
+        /// Asynchronously ends a TextWriter value scope.
+        /// </summary>
+        /// <returns>A task representing the asynchronous operation.</returns>
+        public async Task EndTextWriterValueScopeAsync()
+        {

```

Comment:
Unless I'm wrong, `this.textWriter` is initialized from either `StartTextWriterValueScope` or `StartTextWriterValueScopeAsync`. Should we check that it's not null from this method just in case someone call this method without first calling the method that initializes `this.textWriter`?
If we were to do that, we could make `isWritingJson` a private class member that is initialized from either `StartTextWriterValueScope` or `StartTextWriterValueScopeAsync`.